### PR TITLE
Fix build prefix references in mpi.cfg

### LINKED ISSUE #56: OpenMPI build on osx-arm64 injects unavailable compiler into mpi4py.get_config()
### Solution to issue cannot be found in the documentation.

- [X] I checked the documentation.

### Issue

Example:

```bash
$ conda create -n mpi4py-mpich -c conda-forge mpi4py mpich
$ conda activate mpi4py-mpich
$ python -c "import mpi4py; print(mpi4py.get_config())"
{'mpicc': '/opt/miniconda3/envs/mpi4py-mpich/bin/mpicc', 'mpicxx': '/opt/miniconda3/envs/mpi4py-mpich/bin/mpicxx', 'mpifort': '/opt/miniconda3/envs/mpi4py-mpich/bin/mpifort', 'mpif90': '/opt/miniconda3/envs/mpi4py-mpich/bin/mpif90', 'mpif77': '/opt/miniconda3/envs/mpi4py-mpich/bin/mpif77'}
# all good
```

```bash
$ conda create -n mpi4py-ompi -c conda-forge mpi4py openmpi
$ conda activate mpi4py-ompi
$ python -c "import mpi4py; print(mpi4py.get_config())"
{'mpicc': '/Users/runner/miniforge3/conda-bld/mpi4py_1649058206187/_build_env/bin/mpicc', 'mpicxx': '/Users/runner/miniforge3/conda-bld/mpi4py_1649058206187/_build_env/bin/mpicxx', 'mpifort': '/Users/runner/miniforge3/conda-bld/mpi4py_1649058206187/_build_env/bin/mpifort', 'mpif90': '/Users/runner/miniforge3/conda-bld/mpi4py_1649058206187/_build_env/bin/mpif90', 'mpif77': '/Users/runner/miniforge3/conda-bld/mpi4py_1649058206187/_build_env/bin/mpif77'}
# all paths are from the conda-forge build environment, not the current environment
```

This is only the case on osx-arm64, all other platforms work as intended.

### Installed packages

```shell
# packages in environment at /opt/miniconda3/envs/mpi4py-ompi:
#
# Name                    Version                   Build  Channel
bzip2                     1.0.8                h3422bc3_4    conda-forge
ca-certificates           2022.6.15            h4653dfc_0    conda-forge
libcxx                    14.0.6               h04bba0f_0    conda-forge
libffi                    3.4.2                h3422bc3_5    conda-forge
libgfortran               5.0.0.dev0      11_0_1_hf114ba7_23    conda-forge
libgfortran5              11.0.1.dev0         hf114ba7_23    conda-forge
libzlib                   1.2.12               ha287fd2_2    conda-forge
llvm-openmp               14.0.4               hd125106_0    conda-forge
mpi                       1.0                     openmpi    conda-forge
mpi4py                    3.1.3           py310h2c02159_1    conda-forge
ncurses                   6.3                  h07bb92c_1    conda-forge
openmpi                   4.1.4              h63a3ee9_100    conda-forge
openssl                   3.0.5                h7aea29f_1    conda-forge
pip                       22.2.2             pyhd8ed1ab_0    conda-forge
python                    3.10.5          h4eee789_0_cpython    conda-forge
python_abi                3.10                    2_cp310    conda-forge
readline                  8.1.2                h46ed386_0    conda-forge
setuptools                63.4.3          py310hbe9552e_0    conda-forge
sqlite                    3.39.2               h40dfcc0_0    conda-forge
tk                        8.6.12               he1e0b03_0    conda-forge
tzdata                    2022a                h191b570_0    conda-forge
wheel                     0.37.1             pyhd8ed1ab_0    conda-forge
xz                        5.2.5                h642e427_1    conda-forge
zlib                      1.2.12               ha287fd2_2    conda-forge
```

### Environment info

```shell
active environment : mpi4py-ompi
    active env location : /opt/miniconda3/envs/mpi4py-ompi
            shell level : 2
       user config file : /Users/dion/.condarc
 populated config files : /Users/dion/.condarc
          conda version : 4.12.0
    conda-build version : not installed
         python version : 3.9.12.final.0
       virtual packages : __osx=12.5=0
                          __unix=0=0
                          __archspec=1=arm64
       base environment : /opt/miniconda3  (writable)
      conda av data dir : /opt/miniconda3/etc/conda
  conda av metadata url : None
           channel URLs : https://conda.software.inl.gov/public/osx-arm64
                          https://conda.software.inl.gov/public/noarch
                          https://conda.anaconda.org/conda-forge/osx-arm64
                          https://conda.anaconda.org/conda-forge/noarch
          package cache : /opt/miniconda3/pkgs
                          /Users/dion/.conda/pkgs
       envs directories : /opt/miniconda3/envs
                          /Users/dion/.conda/envs
               platform : osx-arm64
             user-agent : conda/4.12.0 requests/2.27.1 CPython/3.9.12 Darwin/21.6.0 OSX/12.5
                UID:GID : 501:20
             netrc file : None
           offline mode : False
```

### PR DESCRIPTION
**Pull request by minrk**
rewrite build prefix to install prefix in mpi.cfg

closes #56

### CODE REVIEW

**Comment by minrk** [comment]
Looks like the test assumptions don't hold on windows. Will fix that shortly (maybe just skip on Windows?)

File: recipe/run_test.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
+import os
+from subprocess import check_call
+import sys
+
+# openmpi env, before imoprting mpi4py
+os.environ["OMPI_MCA_plm"] = "isolated"
+os.environ["OMPI_MCA_btl_vader_single_copy_mechanism"] = "none"
+os.environ["OMPI_MCA_rmaps_base_oversubscribe"] = "yes"
+os.environ["OMPI_ALLOW_RUN_AS_ROOT"] = "1"
```
Comment (dalcinl) [inline recipe/run_test.py:9]: No, no, this is not going to work. These variables are read by mpiexec, before spawning the actual executable.

File: recipe/run_test.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+os.environ["OMPI_ALLOW_RUN_AS_ROOT_CONFIRM"] = "1"
+
+import mpi4py
```
Comment (dalcinl) [inline recipe/run_test.py:12]: You should not import mpi4py in this script.

**Comment by dalcinl** [comment]
@minrk I don't like your second commit. When using Open MPI, you should not mix testing MPI stuff (like importing mpi4py) with executing stuff using `mpiexec`. Open MPI does not like this "recursive" thing. It may work now, but it may break at any time. I suggest you remove your second commit. If you want to test for the contents of `mpi.cfg`, then better to add an ancillary script you then execute with the environment Python.

**Comment by minrk** [comment]
Fair enough. I split the mpiexec calls into run_test.sh, and left only the Python tests in a separate test file.

**Review by dalcinl — APPROVED** [review]

**Comment by dalcinl** [comment]
LGTM! If you are done with it, please merge.

**Comment by leofang** [comment]
The change LGTM. While I still don't know why we would want to print `mpicc` etc for cross-compiled mpi4py, I have no objection merging this.

**Comment by dalcinl** [comment]
> While I still don't know why we would want to print `mpicc` etc for cross-compiled mpi4py

I get your point, it is less useful in that case. But there may be valid used cases anyway. For example, you can still use the same `mpicc` command to cross-compile other stuff, right?

**Comment by leofang** [comment]
> For example, you can still use the same mpicc command to cross-compile other stuff, right?

No, in this case we'd get a path to osx-arm64 mpicc (from the host env), not the one from the build env (osx-64) used to build the mpi4py. In this case I think you meant to do _native_ compilation, which would be a valid (though likely untested) use case.

**Comment by minrk** [comment]
Yup, it might be useful in user environments, but it will definitely provide the wrong answer if another package uses this value to itself do cross-compilation. In the conda world, `basename($MPICC)` is more likely to be correct in more cases than any absolute path, whereas this record makes more sense in a traditional HPC context where there could easily be several mpi providers, all present on the filesystem at the same time.

When cross-compiling, I think it's ~never a good idea to rely on a dependency's memory of what it was built with without at least the option to override it explicitly when that (totally reasonable) initial guess doesn't hold. In conda, putting `mpi4py` in the build dependencies (like we do with openmpi itself) might work? Cross compiling Python is still a mystery to me.

**Comment by dalcinl** [comment]
> Yup, it might be useful in user environments, but it will definitely provide the wrong answer if another package uses this value to itself do cross-compilation.

Indeed.

> In the conda world, `basename($MPICC)` is more likely to be correct in more cases than any absolute path, whereas this record makes more sense in a traditional HPC context where there could easily be several mpi providers, all present on the filesystem at the same time.

As you can imagine, all this `mpi.cfg` business was added to handle traditional HPC contexts and sloppy users that make a total mess with their `module load ...` invocations. 

I'm wondering whether the mpi4py conda-forge package should store `basename($MPICC)` rather than the absolute path. It should work in most scenarios, but it may also break others.

**Pull request merged**